feat: Add support for macOS app API key notarization

JohnStarich opened the pull request:
* [x] I have read the [contribution documentation](https://github.com/electron/electron-packager/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [code of conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project follows, as appropriate.
* [x] The changes are appropriately documented (if applicable).
* [x] The changes have sufficient test coverage (if applicable).
* [x] The testsuite passes successfully on my local machine (if applicable).

**Summarize your changes:**
Hi all, thanks for this wonderful tool. It's saved me countless hours in developing my app 👏 

This PR adds support for macOS app [API key notarization](https://github.com/electron/electron-notarize/pull/18), which is pretty nice when building a pipeline in open source code. This way I can limit the API key's permissions and completely revoke it if compromised.

The existing notarization validation failed when providing an App Store Connect API key and issuer in place of the expected Apple ID and password combo. This PR loosens it up to accept either pair of credentials.

I came up with a couple questions while writing this up:
* Should this validation be present in electron-packager? Should it instead be deferred to the notarization library?
* Should the failure mode only produce warnings and continue the build without notarization? This seemed odd to me, but perhaps that's normal behavior for electron-packager.

JohnStarich commented:
@malept Thanks for the quick review! I've removed the validation in 8569536, if that's the way we want to go.
After reading through electron-notarize's source, I'm not sure it performs strict validation there today:
https://github.com/electron/electron-notarize/blob/ce868f89895f20bafc3108f1eb0653dc09dcd268/src/index.ts#L40-L49

Perhaps it would be valuable to copy validation from 6b41b29 into Notarize in a follow-up PR?

JohnStarich commented:
Hey @malept, just checking in! Are we waiting on input from @MarshallOfSound?

malept commented:
I think the way forward is:

1. Create a pull request for `electron-notarize` to move the validations from here to there.
2. Get a new release of `electron-notarize`.
3. Update this pull request to update the `electron-notarize` dependency.

JohnStarich commented:
I think the reason validation doesn't exist in Notarize is because it's doing simple translation to CLI args, which would fail only when `notarize` is run (during the actual notarization step).

I'm guessing these validation steps were in Packager to catch failures _before_ running Notarize, since Notarize doesn't have a validation step. I'll investigate adding such a step to Notarize. if you have any suggestions for how that should look, please let me know! Time to learn some TypeScript 😉

JohnStarich commented:
Ok, kicked off a PR on notarize (sorry for all the noise ^^): https://github.com/electron/electron-notarize/pull/35

Also ran the following diff + test successfully with the PR's version of notarize:

```diff
diff --git a/src/mac.js b/src/mac.js
index e9e6575..9329c73 100644
--- a/src/mac.js
+++ b/src/mac.js
@@ -6,7 +6,7 @@ const debug = require('debug')('electron-packager')
 const fs = require('fs-extra')
 const path = require('path')
 const plist = require('plist')
-const { notarize } = require('electron-notarize')
+const { validateAuthorizationArgs, notarize } = require('electron-notarize')
 const { signAsync } = require('electron-osx-sign')
 
 class MacApp extends App {
@@ -396,6 +396,13 @@ function createSignOpts (properties, platform, app, version, notarize, quiet) {
 function createNotarizeOpts (properties, appBundleId, appPath, quiet) {
   const notarizeOpts = properties
 
+  try {
+    validateAuthorizationArgs(notarizeOpts)
+  } catch (e) {
+    common.warning(`Failed validation, notarize will not run: ${e.message}`)
+    return
+  }
+
   // osxNotarize options are handed off to the electron-notarize module, but with a few
   // additions from the main options. The user may think they can pass bundle ID or appPath,
   // but they will be ignored.
diff --git a/test/darwin.js b/test/darwin.js
index afa5db1..51446d8 100644
--- a/test/darwin.js
+++ b/test/darwin.js
@@ -262,6 +262,13 @@ if (!(process.env.CI && process.platform === 'win32')) {
     t.deepEqual(obj.CFBundleURLTypes, expected, 'CFBundleURLTypes did not contain specified protocol schemes and names')
   }))
 
+  test('osxNotarize: missing appleIdPassword', t => {
+    util.setupConsoleWarnSpy()
+    const notarizeOpts = mac.createNotarizeOpts({ appleId: '' })
+    t.falsy(notarizeOpts, 'does not generate options')
+    util.assertWarning(t, 'WARNING: Failed validation, notarize will not run: The appleId property is required when using notarization with appleIdPassword')
+  })
+
   test('osxNotarize: appBundleId not overwritten', t => {
     const args = { appleId: '1', appleIdPassword: '2', appBundleId: 'no' }
     const notarizeOpts = mac.createNotarizeOpts(args, 'yes', 'appPath', true)
```

malept commented:
I've updated the `electron-notarize` dependency on master. Additionally, in #1131 I moved the API docs from a Markdown file to a TypeScript definition file, so you'll have to fix those merge conflicts.

JohnStarich commented:
Thanks for the approval / merge!

Ok, fixed! It seems the bullet points under authentication options were removed in the migration to the new format. I've added them back in my commit ^ but I can definitely remove them if we only want to direct folks to notarize's doc instead 👍

malept commented:
> It seems the bullet points under authentication options were removed in the migration to the new format. I've added them back in my commit ^ but I can definitely remove them if we only want to direct folks to notarize's doc instead 👍

I think I'd rather direct folks to the `electron-notarize` docs.

JohnStarich commented:
Roger that, updated

malept commented on src/mac.js:
```diff
@@ -6,7 +6,7 @@ const debug = require('debug')('electron-packager')
 const path = require('path')
 const plist = require('plist')
-const { notarize } = require('electron-notarize')
+const { validateAuthorizationArgs, notarize } = require('electron-notarize')
```
```suggestion
const { notarize, validateAuthorizationArgs } = require('electron-notarize')
```

malept commented on test/darwin.js:
```diff
@@ -262,18 +262,11 @@ if (!(process.env.CI && process.platform === 'win32')) {
     const notarizeOpts = mac.createNotarizeOpts({ appleId: '' })
     t.falsy(notarizeOpts, 'does not generate options')
-    util.assertWarning(t, 'WARNING: The appleIdPassword sub-property is required when using notarization, notarize will not run')
+    util.assertWarning(t, 'WARNING: Failed validation, notarize will not run: The appleId property is required when using notarization with appleIdPassword')
```
```suggestion
    util.assertWarning(t, 'WARNING: Failed validation, notarization will not run: The appleId property is required when using notarization with appleIdPassword')
```

malept commented on src/mac.js:
```diff
@@ -394,28 +394,19 @@ function createSignOpts (properties, platform, app, version, notarize, quiet) {
+  try {
+    validateAuthorizationArgs(properties)
+  } catch (e) {
+    common.warning(`Failed validation, notarize will not run: ${e.message}`)
```
```suggestion
    common.warning(`Failed validation, notarization will not run: ${e.message}`)
```

JohnStarich commented:
Excellent, thank you! 🚀
